Dropdown for entity attribute selection in Wizard

genthalili opened the pull request:
Hi @mariodavid,

I fixed some errors and implemented new futures on ImportWizard UI : autodetect attributes, select saved configurations (only those related to the selected entity), dropdown proposed entity attribute etc..

this is fixing : #71 , most of what I mentioned in #69 , #31 , #65 , #24 it's a LookUpField
I had almost the same issue as #68 , I fixed it by setting .toLowerCase() comparing with toLowerCase()

Give a try, and let me know what do you think.

Regards,
Gent

mariodavid commented:
Hi @genthalili ,

thanks for your contribution. This is great. I really appreciate that. Before we can merge the different stuff, i'll take a close look.

The configured static analysis tool `codeNarc` will guide you to make the code a little more groovy.
You can run it directly via `./gradlew codenarcMain` or the complete check via `./run-check.sh`

Looking over the code (without going into detail), it is quite some stuff. Perhaps we can enhance it with some tests. I have no tests for the UI parts, only for the business logic.
Normally I use spock for testing (only in the integration tests currently JUnit is used).


I will do a proper code review and let you know about it.

genthalili commented:
Hi @mariodavid ,
Thanks for the tips 👍.  I'll check with `codeNarc` and make the code more groovy!

I'll let you know once I have updated the code.

mariodavid commented on modules/web/src/de/diedavids/cuba/dataimport/web/importwizard/ImportWizard.groovy:
```diff
@@ -99,23 +103,60 @@ class ImportWizard extends AbstractWindow {
+            tabs.each {
+                it.caption = formatMessage(it.name.concat("Title"), "")
+            }
+        } catch (Exception e) {
```
we should not do that - the exception gets lost in space. What can happen an Exception here?

mariodavid commented on modules/web/src/de/diedavids/cuba/dataimport/web/importwizard/ImportWizard.groovy:
```diff
@@ -236,9 +293,13 @@ class ImportWizard extends AbstractWindow {
+            importLogDs.item = importLog
+            toStep5()
+        } catch (Exception i) {
+            showNotification("Couldn't bind attributes, please modify entity attributes and try again.", NotificationType.WARNING_HTML)
```
Unfortunately a lot of stuff can happen when doing the `doDataImport`. It is not necessarily directly related to a binding problem. I'm not sure if the errors should be hidden to the users at that point in the lifetime of the app component. Eventually this is a good idea, but this is pretty beta right now and it gives the user direct intraction points with the developers.

If we want to hide it, we at least need to log it.

genthalili replied:
Same as previous Exception.

I agree. At least log is necessary.

mariodavid commented on modules/web/src/de/diedavids/cuba/dataimport/web/importwizard/import-wizard.xml:
```diff
@@ -74,11 +74,17 @@
+                                 datasource="importConfigurationDs"
+                                 property="entityClass"
+                                 width="100%"/>
+                    <lookupField id="configLookup"
```
I don't fully understand your idea here behind the import configuration. Here's the thing: Either the user decides to use an existing import configuration. Then all of the steps up until the preview are not relevant anymore. They need to be skipped. Because the user does not know what happens if in the entity attribute table will be changed although a import configuration was selected.

So for me it is an either / or approach: Either select a import configuration or configure the import configuration (and potentially store it for later use).

genthalili replied:
I see your point of view.

The solution I proposed is based on the freedom of selecting a configuration or not. First you select the entity to upload, then if there is any configuration in the DB related to that entity (main rule), then this lookup list will show and User will be able to select, otherwise can continue modifying.
But at the same time, the user can select the config and still overwrite it, but the config will not be modified in the DB, which is cool.

To simplify, you can import based on existing config or use the default one which was auto generated.
I think this solution gives more possibilities to the user, adding some message information will be very helpful to the user.

mariodavid replied:
I just tried what you described. You are right. It is possible to select a import config, then this will be loaded. It is adjustable then and with that can be used as a template.

Unfortunately this is pretty unintuitive to understand right now, so I'll not merge it as is.

I'm not sure where this feature of reusing existing import configs is going. The main reason is that the main user of the import wizard. My assumption is that the people that actually want to import the data are normally not the people that are capable of writing groovy scripts.
So most likely the configs will be pre-defined by administrators / developers.
Afterwards they are just "used" by the users. This means that then the easiest path to get from "I want to import" to "I imported successfully" should be preferred. So all the config would need to go out of the wizard.

In this respect your features is just useful for the devs / admins but not for the users, which should be the main user of that screen.

I would suggest that you pull out this part of the PR so that it can get merged more quickly.

We can then in a extra issue & PR talk about that topic in more detail.

genthalili replied:
To be honest the Import Wizard is for devs/admins/master users. It's the same for all tools (MSSQL, SQL dev... ) and the wizard is in the Admin section naturally
Typical user will be asked to use the Import button.. basicly the #32
For admin/dev/master-user : Wizard

Anyway if you think we should avoid config part for now then I created a branch https://github.com/genthalili/cuba-component-data-import/tree/0.4.0-B

You can create remove the code from `ImportWizard` and contribute directly into my repo.

mariodavid replied:
I don't think so. The reason is that I see the current ImportWizard not really as the end of the game. As you might have seen in #76 this is actually my main goal for the users. Not configure anything but still have a wizard with preview, summary etc.

Eventually I think the current Import Wizard can go away, since the normal create operation for an import config is actually quite similar to the current wizard.

I started out with the wizard as a hybrid solution, but I think this is not optimal...

genthalili commented on modules/core/src/de/diedavids/cuba/dataimport/service/ImportWizardServiceBean.groovy:
```diff
@@ -39,4 +41,39 @@ class ImportWizardServiceBean implements ImportWizardService {
+
+    @Override
+    Collection<ImportConfiguration> getImportConfigurations(MetaClass metaClass) {
+        if(metaClass==null)
```
I updated this while fixing the groovy with `codenarc`. Thanks!

genthalili commented on modules/core/src/de/diedavids/cuba/dataimport/service/ImportWizardServiceBean.groovy:
```diff
@@ -39,4 +41,39 @@ class ImportWizardServiceBean implements ImportWizardService {
+
+        def entityName = metaClass.name
+
+        LoadContext loadContext =
```
I didn't see it, that's perfect, I'll definitely use it

genthalili commented on modules/core/src/de/diedavids/cuba/dataimport/service/ImportWizardServiceBean.groovy:
```diff
@@ -39,4 +41,39 @@ class ImportWizardServiceBean implements ImportWizardService {
+    }
+
+    @Override
+    ImportConfiguration getImportConfigurations(MetaClass metaClass, String configName) {
```
True. I had to create a new method accepting `String view` param.

genthalili commented on modules/core/src/de/diedavids/cuba/dataimport/service/ImportWizardServiceBean.groovy:
```diff
@@ -39,4 +41,39 @@ class ImportWizardServiceBean implements ImportWizardService {
+                                .setParameter("cName", configName))
+                        .setView("importConfiguration-view")
+
+        return dataManager.loadList(loadContext).first();
```
I changed it with `getImportConfigurationByName` since this is returning only one `ImportConfiguration`.
Yes, I used `.first()` to get only one. We can update this later based on the use.  
This method is ment to be used from the browser Entity

genthalili commented on modules/global/src/de/diedavids/cuba/dataimport/converter/MetaPropertyMatcher.groovy:
```diff
@@ -11,22 +11,56 @@ class MetaPropertyMatcher {
         result
     }
 
+    List<String> listProperties(List<String> s, String prev, MetaClass selectedEntity) {
```
Sorry for not commenting it :)
This method will return the list of all properties, including **nested** properties.

It's a recursive method looking even for the properties of the property and it's exit condition is : no more property or if becoming "infinity mirror" (Like User and substitue User which actually is a User himself) this is prevented.

Feel free to improve this, so to meat the N:N:1 and N:1 association constrains.

genthalili commented on modules/global/src/de/diedavids/cuba/dataimport/converter/MetaPropertyMatcher.groovy:
```diff
@@ -38,20 +72,24 @@ class MetaPropertyMatcher {
         def mostLikelyMatch = distances.min { it.value }
 
-        if (mostLikelyMatch.value < 5) {
+        if (mostLikelyMatch && mostLikelyMatch.value < 5) {
```
Oh, easy. Thanks!

genthalili commented on modules/global/src/de/diedavids/cuba/dataimport/converter/MetaPropertyMatcher.groovy:
```diff
@@ -38,20 +72,24 @@ class MetaPropertyMatcher {
     private String searchForDirectMatch(List<String> propertiesNames, column) {
         propertiesNames.find {
+            if (it != null) {
+                it = it.toLowerCase();
```
Ok

genthalili commented on modules/web/src/de/diedavids/cuba/dataimport/web/importattributemapper/ImportAttributeMapperEdit.groovy:
```diff
@@ -1,7 +1,81 @@
+    @Inject
+    private LookupField entityAttributeId;
+
+    private List<String> list = new ArrayList<>();
```
Ok, make sense. Fixed!

genthalili commented on modules/web/src/de/diedavids/cuba/dataimport/web/importattributemapper/ImportAttributeMapperEdit.groovy:
```diff
@@ -1,7 +1,81 @@
+                entityAttributeId.setValue(found)
+
+            } catch (Exception e) {
+                showNotification("Something went wrong", NotificationType.ERROR)
```
The exception might happen while getting the `listProperties`. Or the item configuration line 35-36.

But I agree that Exceptions need to be logged somehow... your help is welcomed!

genthalili commented on modules/web/src/de/diedavids/cuba/dataimport/web/importattributemapper/ImportAttributeMapperEdit.groovy:
```diff
@@ -1,7 +1,81 @@
+
+    /* Depreciated */
+
+    void listProperties(String prev, String entityName) {
```
Done!

genthalili commented on modules/web/src/de/diedavids/cuba/dataimport/web/importwizard/ImportWizard.groovy:
```diff
@@ -99,23 +103,60 @@ class ImportWizard extends AbstractWindow {
+
+    }
+
+    void initTitles() {
```
This is to select the config which is coming from ImportConfig Lookup.
User can select one or the "Empty" one, then the default is set as item, to the `importConfigurationDs`.
`defaultImportConfiguration` is the default configuration generated when the view is loaded in `init`

genthalili commented on modules/web/src/de/diedavids/cuba/dataimport/web/importwizard/ImportWizard.groovy:
```diff
@@ -99,23 +103,60 @@ class ImportWizard extends AbstractWindow {
         initImportFileHandler()
         initImportFileParser()
+        initTitles();
+        configLookup.addListener(new ValueListener() {
```
Updated!

genthalili commented on modules/web/src/de/diedavids/cuba/dataimport/web/importattributemapper/import-attribute-mapper-edit.xml:
```diff
@@ -22,7 +22,9 @@
                     <column width="250px">
-                        <field property="entityAttribute"/>
+                        <field property="entityAttribute">
+                            <lookupField id="entityAttributeId" datasource="importAttributeMapperDs" property="entityAttribute"/>
```
Could you explain why not using directly XML ?

mariodavid replied:
Because I think it is the standard way of doing it - I guess? I have actually not seen this solution. 

There are some drawbacks. E.g. I would imagine that with this solution certain features are not directly supported anymore. E.g. when it comes to read-only certain attributes through the security settings.

I would just go the default way as long as it is not required to do otherwise.

genthalili commented:
You should be able to master PR : 
<img width="249" alt="screen shot 2018-05-13 at 18 05 01" src="https://user-images.githubusercontent.com/3913876/39969161-57404de8-56d8-11e8-9e1a-56d8640e8e42.png">

mariodavid commented:
i added some stuff.
Note: the classSize limit is there by purpose. The reason is that otherwise big classes (e.g. the ImportWizard) will get messy and harder to maintain in mid term. So we have to refactor the class and extract parts that are a single concern (like I did with the `MetaPropertyMatcher`)

mariodavid commented on modules/web/src/de/diedavids/cuba/dataimport/web/importwizard/ImportWizard.groovy:
```diff
@@ -144,32 +178,35 @@ class ImportWizard extends AbstractWindow {
-        importConfiguration.booleanTrueValue = 'Yes'
-        importConfiguration.booleanFalseValue = 'No'
-        importConfiguration
+        defaultImportConfiguration = metadata.create(ImportConfiguration)
```
e.g. this part can go into @PostConstruct methods of ImportConfiguration itself --> reduces the size of the wizard

mariodavid commented on modules/web/src/de/diedavids/cuba/dataimport/web/importwizard/ImportWizard.groovy:
```diff
@@ -99,17 +104,46 @@ class ImportWizard extends AbstractWindow {
+    }
+
+    void initConfigLookupChange() {
+        configLookup.addValueChangeListener(new Component.ValueChangeListener() {
```
can be an explicit class `ConfigLookupValueChangeListener` to reduce the class with Class attributes like in this example:
https://github.com/mariodavid/cuba-component-user-inbox/blob/master/modules/web/src/de/diedavids/cuba/userinbox/web/message/ToggleReadActionListener.groovy

mariodavid commented on modules/web/src/de/diedavids/cuba/dataimport/web/importwizard/ImportWizard.groovy:
```diff
@@ -99,17 +104,46 @@ class ImportWizard extends AbstractWindow {
+    ImportConfiguration defaultImportConfiguration
+
+    @Inject
+    Button toStep3
```
you should use the Action, not the button directly

mariodavid commented on modules/global/src/de/diedavids/cuba/dataimport/converter/MetaPropertyMatcher.groovy:
```diff
@@ -8,14 +8,46 @@ import org.springframework.stereotype.Component
+        result
+    }
+
+    List<String> listProperties(List<String> s, String prev, MetaClass selectedEntity) {
```
still the same question as before (it was lost in the history)

mariodavid commented:
I took a look once again over the DropDown solution for the Entity Attribute. As I said before, the way it currently works will break the dynamic attribute functionality, because they are not part of the drop down so it cannot be merged as is.

Also I was re-thinking the solution (of myself) to reference an attribute of an association in order to tell which column should be used for lookup.
I think it would be better if the entity attribute select box only allows direct attributes. In case the selection is a reference, the user has to select in another explicit dropdown the lookup attribute. This will make the problem a little clearer for the user and would hopefully not have created the issues regarding that references could not be bound in the first place...

genthalili commented:
OK, now I understand the idea of dynamic attribute, and I think it really good.
What do you think of this solution : 
When the user checks dynamic attribution then LookuUp changes to simple field ? and unchecked, it becomes a lookuUp with all properties ?
And if you are ok with this solution, should I implement this, or you can do it?

mariodavid commented:
i've partially merged certain stuff from this PR in #76  manually. Let's clear things up and then solve #71  and possibly #24 as well.

Can you merge current master, drop the conflicting implementation of #76? Then i'll do another code review, write down stuff we need to change and implement them here / merge to master before and do slight refactorings.

I really appreciate your effort and really like that you are not bounced backed by the changes I required. So let's finish that thing together with a reduced scope:

1. Back options for Wizards
2. Proper solution for Entity attribute selection

genthalili commented:
I just fixed conflicts, and made it possible to merge with your master after your code review.
I added the view again because I need the configuration and list of attributes from ImportConfiguration :https://github.com/genthalili/cuba-component-data-import/commit/0452e641c1cb758e59dd38de3ec093678dd956ae#diff-120f56c49e1f9df0fb74d711eb2063b4

1. Added possibility to return back : https://github.com/genthalili/cuba-component-data-import/blob/master/modules/web/src/de/diedavids/cuba/dataimport/web/importwizard/ImportWizard.groovy#L275 and it's working
2.1 Check my solution for supporting dynamic attribut  https://github.com/genthalili/cuba-component-data-import/blob/master/modules/web/src/de/diedavids/cuba/dataimport/web/importattributemapper/ImportAttributeMapperEdit.groovy
2.2 and we need to improve this method https://github.com/genthalili/cuba-component-data-import/blob/master/modules/global/src/de/diedavids/cuba/dataimport/converter/MetaPropertyMatcher.groovy#L27-L44 in order to support only associations supported by the component and security lack

Thanks for taking your time on reviewing the code and letting me to contribute on this great project!

genthalili commented:
The version of `declarativecontrollers` should be `0.5.0` to support Cuba `6.8.x`, no ?
https://github.com/genthalili/cuba-component-data-import/blob/master/build.gradle#L102

genthalili commented:
When I'm runing the comentent with my application, I'm getting this message : 
```
2018-05-15 12:31:43.305 ERROR [localhost-startStop-1] com.haulmont.cuba.core.sys.AbstractWebAppContextLoader - Error initializing application
org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'ddcdi$WithImportBrowseAnnotationExecutor': Unsatisfied dependency expressed through field 'buttonsPanelHelper'; nested exception is
org.springframework.beans.factory.NoSuchBeanDefinitionException: No qualifying bean of type 'de.balvi.cuba.declarativecontrollers.web.helper.ButtonsPanelHelper' available: expected at least 1 bean which qualifies as autowire candidate. Dependency annotations: {@javax.inject.Inject()}
 at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor$AutowiredFieldElement.inject(AutowiredAnnotationBeanPostProcessor.java:588) ~[spring-beans-4.3.12.RELEASE.jar:4.3.12.RELEASE]
 at 
```
But when running directly from data-import component it works correctly, do you think there something wrong with dependencies ?

mariodavid commented:
I had the same issue in the component itself. I'm not really sure what is wrong with the config of the app component.

Ive started to do the implementation for attribute selection i talked about.
It is not fully working but fro a UI perspective this the desired goal, because it eliminates different problem for users all together. Will probably a breaking change though...

genthalili commented:
Hi @mariodavid,
I check the new updated you made, and it looks very good now.
Two feedbacks form my side : 
1. "." path fileColumn is not supported for eventual associate columns, should we implement one to make this possible? (i.e. group.something => if ASSOC attribute existes then it will select "group" and look for a "something" property inside group, if DIRECT attribute will select "something" then search for "group") ==> I propose a solution with my last commit, what do you think?
2.  my input column alias as "lastname" is selecting "name" because of casesensitive in `searchForDirectMatch` so then is going to `findNearestMatchIfPossible` which returns "name" instead of "**lastName**"
3. I think there should be something to prevent selecting same attribute for different column name

mariodavid commented on modules/web/src/de/diedavids/cuba/dataimport/web/importattributemapper/EntityAttributeSelectionHandler.groovy:
```diff
@@ -104,7 +104,8 @@ class EntityAttributeSelectionHandler {
         if (metaProperty) {
             MetaClass importConfigurationMetaClass = metadata.session.getClass(importAttributeMapperDs.item.configuration.entityClass)
-            associationLookupAttribute.optionsMap = metadataSelector.getLookupMetaProperties(importConfigurationMetaClass.getProperty(metaProperty).domain.properties)
+            def assocProperties = importConfigurationMetaClass.getProperty(metaProperty)?.range?.asClass()?.ownProperties
```
this is good. I just experienced that on my own. Should work...

mariodavid commented on modules/global/src/de/diedavids/cuba/dataimport/converter/MetaPropertyMatcher.groovy:
```diff
@@ -3,24 +3,44 @@ package de.diedavids.cuba.dataimport.converter
-        String result = ''
-        if (possibleProperty && isSimpleDatatype(possibleProperty)) {
-                result = possibleProperty.name
+    ImportAttributeMapper findEntityAttributeForColumn(Integer index, String column, MetaClass selectedEntity) {
```
i think you mixed up the responsibilities here a little bit. From the naming: `ImportAttributeMapperCreator` is for creating the mappers. `MetaPropertyMatcher` is for matching, not for creating. I think the logic belongs to the former one.
Ill change that and add some tests, because there are some errors.

genthalili replied:
Now that you mentioned, I see, I should have done this is `ImportAttributeMapperCreator`, sorry about that.
I agree the complicity is a bit limit here, need to be crated outside this method

mariodavid commented on modules/global/src/de/diedavids/cuba/dataimport/converter/MetaPropertyMatcher.groovy:
```diff
@@ -3,24 +3,44 @@ package de.diedavids.cuba.dataimport.converter
+        if (possibleProperty && !isSimpleDatatype(possibleProperty)) {
+            MetaClass assocAttribute = possibleProperty.range?.asClass()
+            if (possiblePropertyName && assocAttribute) {
+                result.associationLookupAttribute = findPropertyByColumn(assocAttribute, inputColumnList[(1)])?.name
```
I got a `ArrayIndexOutOfBounds` because inputColumnsList might have only one entry (as in my case).

As well -> I will add some tests around it and nail the features by test cases. Otherwise this bean becomes a beast to maintain...

genthalili commented:
@mariodavid  After new updates (Attribute binding) ASSOCIATION is not working, I'm getting `Entity validation failed`  caused by `com.haulmont.cuba.core.app.importexport.EntityImportExport#importEntities(java.util.Collection<? extends com.haulmont.cuba.core.entity.Entity>, com.haulmont.cuba.core.app.importexport.EntityImportView, boolean)`
So the FK value is in `commitContext.getCommitInstances()` line 247 of EntityImportExport.java

Have you tried associations ?

mariodavid commented:
yes, i had a fix for that locally, just pushed it...
